File: 승준/20230624/complex_number_2667.java
import java.io.*;
import java.util.*;

public class complex_number_2667 {
    static int n;
    static int[][] graph;
    static int number;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    static int count;
    static Queue<int[]> queue;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        graph = new int[n][n];
        number = 1;
        count = 0;
        queue = new ArrayDeque<>();

        for (int i = 0; i < n; i++) {
            String row = br.readLine();
            for (int j = 0; j < n; j++) {
                graph[i][j] = Character.getNumericValue(row.charAt(j));
            }
        }

        int[] countArr = new int[n * n];
        int islandCount = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (graph[i][j] == 1) {
                    number++;
                    count = bfs(number, i, j);
                    countArr[islandCount++] = count;
                }
            }
        }

        System.out.println(number-1);
        int[] nonZeroCountArr = Arrays.stream(countArr).filter(x -> x != 0).toArray();
        Arrays.sort(nonZeroCountArr);
        for (int i = 0; i < islandCount; i++) {
            System.out.println(nonZeroCountArr[i]);
        }
    }

    static int bfs(int number, int sy, int sx) {
        queue.clear();
        queue.add(new int[]{sy, sx});
        graph[sy][sx] = number;
        count = 1;

        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            int y = curr[0];
            int x = curr[1];

            for (int d = 0; d < 4; d++) {
                int ny = y + dy[d];
                int nx = x + dx[d];
                if ( 0 <= ny && ny < n && 0 <= nx && nx < n && graph[ny][nx] == 1) {
                    queue.add(new int[]{ny, nx});
                    graph[ny][nx] = number;
                    count++;
                }
            }
        }
        return count;
    }
}

File: 승준/20230624/complex_number_2667.js
const readline = require("readline");

input = [];
let n;
let graph;
let number = 1;
const dx = [-1, 1, 0, 0];
const dy = [0, 0, -1, 1];
let count;
let queue;

const bfs = (number, sy, sx) => {
  queue = [];
  queue.push([sy, sx]);
  graph[sy][sx] = number;
  count = 1;

  while (queue.length > 0) {
    const [y, x] = queue.shift();

    for (let d = 0; d < 4; d++) {
      const ny = y + dy[d];
      const nx = x + dx[d];
      if (0 <= ny && ny < n && 0 <= nx && nx < n && graph[ny][nx] === 1) {
        queue.push([ny, nx]);
        graph[ny][nx] = number;
        count++;
      }
    }
  }

  return count;
};

readline
  .createInterface(process.stdin, process.stdout)
  .on("line", (line) => {
    input.push(line);
  })
  .on("close", () => {
    n = parseInt(input[0]);
    graph = new Array(n);

    for (let i = 1; i < n + 1; i++) {
      graph[i - 1] = input[i].split("").map(Number);
    }

    const countArr = [];
    let islandCount = 0;

    for (let i = 0; i < n; i++) {
      for (let j = 0; j < n; j++) {
        if (graph[i][j] === 1) {
          number++;
          count = bfs(number, i, j);
          countArr.push(count);
          islandCount++;
        }
      }
    }

    console.log(number - 1);
    countArr.sort((a, b) => a - b);
    for (let i = 0; i < islandCount; i++) {
      console.log(countArr[i]);
    }

    process.exit();
  });
